File: day3.py
from os import O_TEXT


def ParseFile(path: str) -> list[str]:
    res = []
    with open(path) as file:
        for line in file:
            res.append(line.strip())
    return res

def part1(path: str) -> int:
    nums = ParseFile(path)
    bits = len(nums[0])

    gamma = ""
    epsilon = ""

    for i in range(bits):
        count = 0
        for num in nums:
            if num[i] == '1':
                count += 1
        if count > len(nums) / 2:
            gamma += '1'
            epsilon += '0'
        else:
            gamma += '0'
            epsilon += '1'
    
    return int(gamma, 2) * int(epsilon, 2)

def part2(path: str) -> int:
    nums = ParseFile(path)
    bits = len(nums[0])

    oxygen = nums
    co2 = nums

    for i in range(bits):
        if len(oxygen) <= 1:
            break

        # Calculate set bits in ith position
        count = 0
        for num in oxygen:
            if num[i] == '1':
                count += 1

        # Erase the least common bit (0 or 1)
        if count >= len(oxygen) / 2:
            oxygen = [x for x in oxygen if x[i] == '1']
        else:
            oxygen = [x for x in oxygen if x[i] == '0']

    for i in range(bits):
        if len(co2) <= 1:
            break

        # Calculate set bits in ith position
        count = 0
        for num in co2:
            if num[i] == '1':
                count += 1
                
        # Erase the most common bit (0 or 1)
        if count < len(co2) / 2:
            co2 = [x for x in co2 if x[i] == '1']
        else:
            co2 = [x for x in co2 if x[i] == '0']

    return int(oxygen[0], 2) * int(co2[0], 2)


if __name__ == '__main__':
    print(f"PART 1: {part1('input/day3.txt')}")
    print(f"PART 2: {part2('input/day3.txt')}")

File: day3.cpp
#include "utilities.h"

int part1(char* path)
{
    std::ifstream file(path);
    const auto numbers = ReadStrings(file);

    int bits = numbers.front().size();
    int mask = (1 << bits) - 1;
    int gamma = 0;

    for (int i = 0; i < bits; i++) {
        int cnt = 0;
        for (auto& num : numbers)
            cnt += num[i] - '0';
        gamma = (gamma << 1) | (cnt > numbers.size() >> 1);
    }

    return gamma * (~gamma & mask);
}

int part2(char* path)
{
    std::ifstream file(path);
    const auto numbers = ReadStrings(file);
    int bits = numbers.front().size();

    auto g = numbers;
    for (int i = 0; i < bits && g.size() > 1; i++) {
        int cnt = 0;
        for (auto& num : g)
            cnt += num[i] - '0';
        char c = '0' + (2 * cnt >= g.size());
        g.erase(std::remove_if(g.begin(), g.end(), [i, c](auto& s) { return s[i] != c; }), g.end());
    }

    auto e = numbers;
    for (int i = 0; i < bits && e.size() > 1; i++) {
        int cnt = 0;
        for (auto& num : e)
            cnt += num[i] - '0';
        char c = '0' + (2 * cnt < e.size());
        e.erase(std::remove_if(e.begin(), e.end(), [i, c](auto& s) { return s[i] != c; }), e.end());
    }

    return std::stoi(g.front(), nullptr, 2) * std::stoi(e.front(), nullptr, 2);
}

int main(int argc, char** argv)
{
    char* path;

    if(argc != 2)
    {
        path = new char[100];
        sprintf(path, "input/%s.txt", argv[0]);
    }
    else path = argv[1];

    printf("PART 1: %i\n", part1(path));
    printf("PART 2: %i\n", part2(path));

    if(argc != 2)
        delete[] path;
}